### PR TITLE
Add object detection + segmentation transforms

### PR DESCRIPTION
**Pull request by amyeroberts**
# What does this PR do?

Adds logic for processing bounding boxes and some additional transforms (`rgb_to_id`, `id_to_rgb`) needed for DETR. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: src/transformers/image_transforms.py [hunk 1]
```diff
@@ -420,3 +426,106 @@ def center_crop(
...
+    Converts RGB color to unique ID.
+    """
+    if isinstance(color, np.ndarray) and len(color.shape) == 3:
+        if color.dtype == np.uint8:
+            color = color.astype(np.int32)
+        return color[:, :, 0] + 256 * color[:, :, 1] + 256 * 256 * color[:, :, 2]
+    return int(color[0] + 256 * color[1] + 256 * 256 * color[2])
+
+
+def id_to_rgb(id_map):
```
Comment (NielsRogge) [inline src/transformers/image_transforms.py:515]: Same here.

File: src/transformers/image_transforms.py [hunk 2]
```diff
@@ -420,3 +426,106 @@ def center_crop(
...
+    elif is_tf_tensor(bboxes_corners):
+        return _corners_to_center_format_tf(bboxes_corners)
+
+    raise ValueError(f"Unsupported input type {type(bboxes_corners)}")
+
+
+# 2 functions below copied from https://github.com/cocodataset/panopticapi/blob/master/panopticapi/utils.py
+# Copyright (c) 2018, Alexander Kirillov
+# All rights reserved.
+def rgb_to_id(color):
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/image_transforms.py:504]: Yes - I was bundling together functionality for DETR and quickly wrote a draft PR title :)

File: src/transformers/image_transforms.py [hunk 3]
```diff
@@ -420,3 +426,106 @@ def center_crop(
...
+
+
+def _center_to_corners_format_numpy(bboxes_center: np.ndarray) -> np.ndarray:
+    x_c, y_c, w, h = bboxes_center.T
+    bboxes_corners = np.stack([x_c - 0.5 * w, y_c - 0.5 * h, x_c + 0.5 * w, y_c + 0.5 * h], axis=-1)
+    return bboxes_corners
+
+
+def _center_to_corners_format_tf(bboxes_center: "tf.Tensor") -> "tf.Tensor":
+    x_c, y_c, w, h = tf.unstack(bboxes_center, axis=-1)
```
Comment (sgugger) [inline src/transformers/image_transforms.py:444]: Let's use real variable names pretty please :-)

File: src/transformers/image_transforms.py [hunk 4]
```diff
@@ -420,3 +426,106 @@ def center_crop(
...
+    return torch.stack(b, dim=-1)
+
+
+def _corners_to_center_format_numpy(bboxes_corners: np.ndarray) -> np.ndarray:
+    x_0, y_0, x_1, y_1 = bboxes_corners.T
+    bboxes_center = np.stack([(x_0 + x_1) / 2, (y_0 + y_1) / 2, (x_1 - x_0), (y_1 - y_0)], axis=-1)
+    return bboxes_center
+
+
+def _corners_to_center_format_tf(bboxes_corners: "tf.Tensor") -> "tf.Tensor":
```
Comment (sgugger) [inline src/transformers/image_transforms.py:479]: Same there

File: src/transformers/image_transforms.py [hunk 5]
```diff
@@ -420,3 +426,106 @@ def center_crop(
...
+    x_c, y_c, w, h = tf.unstack(bboxes_center, axis=-1)
+    bboxes_corners = tf.stack([x_c - 0.5 * w, y_c - 0.5 * h, x_c + 0.5 * w, y_c + 0.5 * h], axis=-1)
+    return bboxes_corners
+
+
+# 2 functions below inspired by https://github.com/facebookresearch/detr/blob/master/util/box_ops.py
+def center_to_corners_format(bboxes_center: TensorType) -> TensorType:
+    """
+    Converts bounding boxes from center format (center_x, center_y, width, height) to corners format (x_0, y_1, x_1,
+    y_1).
```
Comment (sgugger) [inline src/transformers/image_transforms.py:453]:
```suggestion
    Converts bounding boxes from center format (center_x, center_y, width, height) to corners format `(left, top, right, bottom)`.
```
This is clearer to me than x_0/y_1

File: src/transformers/image_transforms.py [hunk 6]
```diff
@@ -420,3 +426,106 @@ def center_crop(
...
+
+def _corners_to_center_format_tf(bboxes_corners: "tf.Tensor") -> "tf.Tensor":
+    x_0, y_0, x_1, y_1 = tf.unstack(bboxes_corners, axis=-1)
+    bboxes_center = tf.stack([(x_0 + x_1) / 2, (y_0 + y_1) / 2, (x_1 - x_0), (y_1 - y_0)], axis=-1)
+    return bboxes_center
+
+
+def corners_to_center_format(bboxes_corners: TensorType) -> TensorType:
+    """
+    Converts bounding boxes from corners format (x_0, y_1, x_1, y_1) to center format (center_x, center_y, width,
```
Comment (sgugger) [inline src/transformers/image_transforms.py:487]: Same here for the docstrings.

**Review by sgugger — APPROVED** [review]
Nice to add those utils. Now to use them everywhere ;-)

**Review by alaradirik — APPROVED** [review]
Thanks for the PR! I agree with Niels and Sylvain's comments but looks good to me otherwise.

File: src/transformers/image_transforms.py [hunk 7]
```diff
@@ -420,3 +424,147 @@ def center_crop(
...
+    )
+    return bboxes_corners
+
+
+# 2 functions below inspired by https://github.com/facebookresearch/detr/blob/master/util/box_ops.py
+def center_to_corners_format(bboxes_center: TensorType) -> TensorType:
+    """
+    Converts bounding boxes from center format to corners format.
+
+    center format: contains the coordinate for the center of the box and its the width, height dimensions
```
Comment (amyeroberts) [inline src/transformers/image_transforms.py:464]:
```suggestion
    center format: contains the coordinate for the center of the box and its width, height dimensions
```

File: src/transformers/image_transforms.py (hunk 6 above)
Comment (amyeroberts) [reply src/transformers/image_transforms.py]: I've updated the vars and docstrings. LMK if the updates are OK :)

**Review by sgugger — APPROVED** [review]
Thanks for iterating!

**Pull request merged**